src/transformers/models/ict/modeling_ict.py

```diff
@@ -0,0 +1,837 @@
+        self.guided_upsampler = IctGuidedUpsampler(config)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.transformer.embeddings.token_embedding
+
+    @add_start_docstrings_to_model_forward(ICT_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=MaskedImageModelingOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        pixel_values: Optional[torch.Tensor],
+        original_images: ImageInput,
+        bool_masked_pos: Optional[torch.BoolTensor] = None,
+        clusters: Optional[np.ndarray] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutput]:
+        r"""
+        Returns:
+
+        Example:
+        ```python
+        >>> import torch
+        >>> import numpy as np
+        >>> from PIL import Image
+        >>> import requests
+
+        >>> from transformers import AutoImageProcessor, IctModel
+
+        >>> image_processor = image_AutoImageProcessor.from_pretrained("sheonhan/ict-imagenet-256")
+        >>> model = IctModel.from_pretrained("sheonhan/ict-imagenet-256")
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> pixel_values = image_processor(image, return_tensors="pt").pixel_values
+        >>> clusters = image_processor.clusters
+
+        >>> # create random boolean mask of shape (batch_size, num_patches)
+        >>> bool_masked_pos = torch.randint(low=0, high=2, size=(pixel_values.shape[0], pixel_values.shape[1])).bool()
+
+        >>> outputs = model(pixel_values, bool_masked_pos=bool_masked_pos, clusters=clusters)
+        ```"""
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+        if pixel_values is None:
+            raise ValueError("You have to specify pixel_values")
+
+        outputs = self.transformer(
+            pixel_values,
+            bool_masked_pos=bool_masked_pos,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+        sequence_output = outputs[0]
+        batch_size, sequence_length, num_channels = sequence_output.shape
+        height = width = math.floor(sequence_length**0.5)
+        sequence_output = sequence_output.permute(0, 2, 1).reshape(batch_size, num_channels, height, width)
+
+        # modified from https://github.com/raywzy/ICT/blob/59dd12d374d47cdf0dce90923017ca3657e6aa0b/Transformer/inference.py#L107-L126
+        clusters = np.rint(127.5 * (clusters + 1.0))
+        clusters = torch.from_numpy(clusters)
+        recovered_pixel_values = [clusters[pixel_values[i]].view(height, width, 3) for i in range(batch_size)]
```

Comment:
If I follow the way things are done in the original code, I first have to recover the quantized vectors to an image and use _that_ as input to the Guided Sampler. So we need to have `clusters` which seems somewhat wonky to me...? 

@amyeroberts I noticed that [ImageGPT uses `clusters` from ImageProcessor](https://github.com/huggingface/transformers/blob/main/src/transformers/models/imagegpt/modeling_imagegpt.py#L992), so in the test code, I followed the same pattern — using `clusters` from the processor as input to this forward method

Reply:
Also, where does `generate` method here come from? https://github.com/huggingface/transformers/blob/main/src/transformers/models/imagegpt/modeling_imagegpt.py#L988